### PR TITLE
[10.x] Fix breaking changes in after validation

### LINKED ISSUE #49863: Before/After validation rules ignore the validation state of the dependent field
### Laravel Version

10.34.2

### PHP Version

8.3.2

### Database Driver & Version

_No response_

### Description

Validator's `before` and `after` rules are trying to construct the second operand from non-validated potentially invalid data ([ValidatesAttributes::compareDates()](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Validation/Concerns/ValidatesAttributes.php#L256) is retrieving the data using [Validator::getValue()](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Validation/Validator.php#L1125) which returns raw data). This causes unexpected exceptions.

### Steps To Reproduce

**Sample script**
```php
<?php

require_once(__DIR__ . '/vendor/autoload.php');
$app = require __DIR__.'/bootstrap/app.php';
$app->make(Illuminate\Contracts\Console\Kernel::class)->bootstrap();

use Illuminate\Support\Facades\Validator;

$data = [
    'dt1' => ['this' => ['is' => ['a' => ['mess']]]],
    'dt2' => 'some invalid date',
];

$val = Validator::make($data, [
    'dt1' => 'required|date',
    'dt2' => 'required|date|before:dt1',
]);

$val->validate();
```

**Output**
```

   TypeError 

  DateTime::__construct(): Argument #1 ($datetime) must be of type string, array given

  at vendor/nesbot/carbon/src/Carbon/Traits/Creator.php:89
     85▕             setlocale(LC_NUMERIC, 'C'); // @codeCoverageIgnore
     86▕         }
     87▕ 
     88▕         try {
  ➜  89▕             parent::__construct($time ?: 'now', static::safeCreateDateTimeZone($tz) ?: null);
     90▕         } catch (Exception $exception) {
     91▕             throw new InvalidFormatException($exception->getMessage(), 0, $exception);
     92▕         }
     93▕ 

      +13 vendor frames 

  14  poc.php:19
      Illuminate\Validation\Validator::validate()
```

**Expected result**

Validation exception stating that both values are invalid.

**Actual result**

TypeError fatal error.

### PR DESCRIPTION
**Pull request by MrPunyapal**
I had a [PR #49871](https://github.com/laravel/framework/pull/49871) to fix [Issue #49863](https://github.com/laravel/framework/issues/49863) but from this https://github.com/laravel/framework/pull/49871#issuecomment-1921544394 conversation it looks breaking change. also solving [Issue #49955](https://github.com/laravel/framework/issues/49955) 

if it is a breaking change then this PR will solve it. added tests for it too. 

```php
$v = new Validator($trans, ['y' => '1970-01-01'], ['x' => 'nullable|date', 'y' => 'nullable|date|after:x']);
$this->assertTrue($v->passes());

$v = new Validator($trans, ['dates' => [['start_at' => '2024-02-02 12:00:00', 'ends_at' => '2024-02-02 12:00:00']]], ['dates.*.start_at' => 'date', 'dates.*.ends_at' => 'date|after:start_at']);
$this->assertTrue($v->passes());

$v = new Validator($trans, ['date_end' => '2021-09-17 13:28:47'], ['date_start' => 'nullable|date', 'date_end' => 'nullable|date|after_or_equal:date_start']);
$this->assertTrue($v->passes());
```

>[!NOTE]
> last 2 tests were not passing with `x`, `y` at place of `starts_at` or `date_start` and `ends_at` or `date_end`
> even if I revert my previous  [PR #49871](https://github.com/laravel/framework/pull/49871) changes.
> we can look into that later

### CODE REVIEW

**Comment by MrPunyapal** [comment]
@calebdw

File: src/Illuminate/Validation/Concerns/ValidatesAttributes.php [hunk 1]
```diff
@@ -255,7 +255,7 @@ protected function compareDates($attribute, $value, $parameters, $operator)
         if (is_null($date = $this->getDateTimestamp($parameters[0]))) {
             $comparedValue = $this->getValue($parameters[0]);
 
-            if (! is_string($comparedValue) && ! is_numeric($comparedValue) && ! $comparedValue instanceof DateTimeInterface) {
+            if (! is_null($comparedValue) && ! is_string($comparedValue) && ! is_numeric($comparedValue) && ! $comparedValue instanceof DateTimeInterface) {
```
Comment (calebdw) [inline src/Illuminate/Validation/Concerns/ValidatesAttributes.php:258]: This works, although the latter conditions are repeated above  as well, so I would probably extract a method:

```suggestion
            if (! is_null($comparedValue) && ! $this->hasDateTypes($comparedValue)) {
```

File: tests/Validation/ValidationValidatorTest.php [hunk 2]
```diff
@@ -5986,6 +5986,17 @@ public function testBeforeAndAfter()
 
         $v = new Validator($trans, ['x' => ['a' => ['v' => 'c']], 'y' => 'invalid'], ['x' => 'date', 'y' => 'date|before:x']);
         $this->assertTrue($v->fails());
+
+        $v = new Validator($trans, ['x' => '1970-01-01'], ['x' => 'nullable|date', 'y' => 'nullable|date|after:x']);
+        $this->assertTrue($v->passes());
```
Comment (calebdw) [inline tests/Validation/ValidationValidatorTest.php:5991]: You're missing a test here (where `x` is explicitly `null`):

```php
$v = new Validator($trans, ['x' => null', 'y' => '1970-01-01'], ['x' => 'nullable|date', 'y' => 'nullable|date|after:x']);
$this->assertTrue($v->passes());
```

**Review by calebdw — COMMENTED** [review]
This fixes my issue, just added some comments

**Comment by MrPunyapal** [comment]
it has some weird thing 

> This fixes my issue, just added some comments

I don't think your issue is fixed?

**Comment by calebdw** [comment]
Yes, adding `! is_null(...)` to the conditions allows my validation test to pass

**Comment by MrPunyapal** [comment]
> Yes, adding `! is_null(...)` to the conditions allows my validation test to pass

but that test case is failing?

File: tests/Validation/ValidationValidatorTest.php [hunk 3]
```diff
@@ -5986,6 +5986,20 @@ public function testBeforeAndAfter()
 
         $v = new Validator($trans, ['x' => ['a' => ['v' => 'c']], 'y' => 'invalid'], ['x' => 'date', 'y' => 'date|before:x']);
         $this->assertTrue($v->fails());
+
+        $v = new Validator($trans, ['y' => '1970-01-01'], ['x' => 'nullable|date', 'y' => 'nullable|date|after:x']);
+        $this->assertTrue($v->passes());
```
Comment (calebdw) [inline tests/Validation/ValidationValidatorTest.php:5991]: > but that test case is failing?

No, this is the test case that is failing (when `x` is missing from the array). This tests before and after the changes introduced by your PR.

Comment (calebdw) [reply tests/Validation/ValidationValidatorTest.php:5991]: Ah, I found the issue, when compared against `null`, it's comparing against the current time (`now`):

This should properly capture the behaviour before and after the PR:

```php
        $v = new Validator($trans, ['y' => '1970-01-01'], ['x' => 'nullable|date', 'y' => 'nullable|date|before:x']);
        $this->assertTrue($v->passes());

        $v = new Validator($trans, ['y' => now()->addDay()], ['x' => 'nullable|date', 'y' => 'nullable|date|after:x']);
        $this->assertTrue($v->passes());

        $v = new Validator($trans, ['x' => null, 'y' => '1970-01-01'], ['x' => 'nullable|date', 'y' => 'nullable|date|before:x']);
        $this->assertTrue($v->passes());

        $v = new Validator($trans, ['x' => null, 'y' => now()->addDay()], ['x' => 'nullable|date', 'y' => 'nullable|date|after:x']);
        $this->assertTrue($v->passes());

```

File: src/Illuminate/Validation/Concerns/ValidatesAttributes.php [hunk 4]
```diff
@@ -255,6 +255,10 @@ protected function compareDates($attribute, $value, $parameters, $operator)
         if (is_null($date = $this->getDateTimestamp($parameters[0]))) {
             $comparedValue = $this->getValue($parameters[0]);
 
+            if (is_null($comparedValue)) {
+                return true;
+            }
```
Comment (calebdw) [inline src/Illuminate/Validation/Concerns/ValidatesAttributes.php:260]: this change breaks existing behaviour

File: tests/Validation/ValidationValidatorTest.php (hunk 3 above)
Comment (MrPunyapal) [reply tests/Validation/ValidationValidatorTest.php:5991]: Is it a different bug? 
not a breaking change?

Comment (calebdw) [reply tests/Validation/ValidationValidatorTest.php:5991]: not a breaking change

Comment (MrPunyapal) [reply tests/Validation/ValidationValidatorTest.php:5991]: lemme just remove that test case for now.

and solve that bug with separate PR?

Comment (calebdw) [reply tests/Validation/ValidationValidatorTest.php]: I don't see any bug...

**Comment by iamgergo** [comment]
Might be related: https://github.com/laravel/framework/pull/49668

File: tests/Validation/ValidationValidatorTest.php (hunk 3 above)
Comment (MrPunyapal) [reply tests/Validation/ValidationValidatorTest.php]: this test case was event failing before my PR
```php
$v = new Validator($trans, ['dates' => [['x' => '2024-02-02 12:00:00', 'y' => '2024-02-02 12:00:00']]], ['dates.*.x' => 'date', 'dates.*.y' => 'date|after:x']);
$this->assertTrue($v->passes());
```

but I just replace x with starts_at and y with ends_at this passes

```php
$v = new Validator($trans, ['dates' => [['start_at' => '2024-02-02 12:00:00', 'ends_at' => '2024-02-02 12:00:00']]], ['dates.*.start_at' => 'date', 'dates.*.ends_at' => 'date|after:start_at']);
$this->assertTrue($v->passes());
```

File: tests/Validation/ValidationValidatorTest.php (hunk 2 above)
Comment (MrPunyapal) [reply tests/Validation/ValidationValidatorTest.php:5991]: this was also not passing before my  [PR #49871](https://github.com/laravel/framework/pull/49871)

**Pull request closed without merging**